Runtime defaults config supplied via `RunTimeConfigurationDefaultBuildItem` are not overridden on restart during dev mode

## Describe the bug

An extension author can provide default values via the `RunTimeConfigurationDefaultBuildItem`, this is the case for dev services. In dev mode, the same key can have different values on each restart e.g a new container was spun by testcontainers, taking a new random/different port. The behaviour I am seeing with dev services is that a new restart of the container having a different port does not update the config value and the first supplied value is always cached.  

### Expected behavior

Defaults values supplied via `RunTimeConfigurationDefaultBuildItem` should always be taken into account on each app restart in dev mode and not cache the first value. 

### Actual behavior

The first default value of a config item is cached and subsequent overrides of the values are not taken into account. 

This is the case for datasource devservices and mongo devservices.  This leads to connection failures. 

## To Reproduce

The issue can be reproduced with the current quickstarts e.g the [mongodb one](https://github.com/quarkusio/quarkus-quickstarts/tree/development/mongodb-quickstart)

 
Steps to reproduce the behavior:
1. Using the `https://github.com/quarkusio/quarkus-quickstarts/tree/development/mongodb-quickstart` quickstart 
2. Open the `vim src/main/resources/application.properties` configuration file and add the following entry only
```properties
quarkus.mongodb.devservices.image-name=mongo
```
while making sure that the connection-string property is commented out or removed.
3. start the quickstart in dev mode with `mvn quarkus:dev`
4. visits the `http://localhost:8080/q/dev/io.quarkus.quarkus-vertx-http/config` page and observe the auto-configured connection string. 
Also observe the output of the `docker ps` command to see the current mongo container mapped port. This should match the port in the config page. 
5. Open the `vim src/main/resources/application.properties` configuration file and change the image tag to `mongo:4.4.5` this will lead to a new container being spin up. 
```properties
quarkus.mongodb.devservices.image-name=mongo:4.4.5
```
while making sure that the connection-string property is commented out or removed.
6. Observe the logs of the command in (3) and you'll see logs like below
```log
(cluster-ClusterId{value='6095541bbe5033399414ba51', description='null'}-localhost:49194) Exception in monitor thread while connecting to server localhost:49194: com.mongodb.MongoSocketOpenException: Exception opening socket
	at com.mongodb.internal.connection.AsynchronousSocketChannelStream$OpenCompletionHandler.failed(AsynchronousSocketChannelStream.java:124)
	at java.base/sun.nio.ch.Invoker.invokeUnchecked(Invoker.java:129)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finishConnect(UnixAsynchronousSocketChannelImpl.java:283)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finish(UnixAsynchronousSocketChannelImpl.java:198)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.onEvent(UnixAsynchronousSocketChannelImpl.java:213)
	at java.base/sun.nio.ch.EPollPort$EventHandlerTask.run(EPollPort.java:306)
	at java.base/sun.nio.ch.AsynchronousChannelGroupImpl$1.run(AsynchronousChannelGroupImpl.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.finishConnect(UnixAsynchronousSocketChannelImpl.java:252)
	... 7 more
```

Observing the output from `docker ps` again will show that a new container was started by devservices but the app is using the old container which can be observed in the config page.

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java
@@ -273,7 +273,7 @@ public final class RunTimeConfigurationGenerator {
     }
 
     public static final class GenerateOperation implements AutoCloseable {
-        final boolean devMode;
+        final boolean liveReloadPossible;
         final LaunchMode launchMode;
         final AccessorFinder accessorFinder;
         final ClassOutput classOutput;
@@ -291,8 +291,6 @@ public final class RunTimeConfigurationGenerator {
         // default values given in the build configuration
         final Map<String, String> specifiedRunTimeDefaultValues;
         final Map<String, String> buildTimeRunTimeVisibleValues;
-        // default values produced by extensions via build item
-        final Map<String, String> runTimeDefaults;
         final Map<Container, MethodDescriptor> enclosingMemberMethods = new HashMap<>();
         final Map<Class<?>, MethodDescriptor> groupInitMethods = new HashMap<>();
         final Map<Class<?>, FieldDescriptor> configRootsByType = new HashMap<>();
@@ -325,7 +323,7 @@ public final class RunTimeConfigurationGenerator {
 
         GenerateOperation(Builder builder) {
             this.launchMode = builder.launchMode;
-            this.devMode = builder.launchMode == LaunchMode.DEVELOPMENT;
+            this.liveReloadPossible = builder.liveReloadPossible;
             final BuildTimeConfigurationReader.ReadResult buildTimeReadResult = builder.buildTimeReadResult;
             buildTimeConfigResult = Assert.checkNotNullParam("buildTimeReadResult", buildTimeReadResult);
             specifiedRunTimeDefaultValues = Assert.checkNotNullParam("specifiedRunTimeDefaultValues",
@@ -334,7 +332,6 @@ public final class RunTimeConfigurationGenerator {
                     buildTimeReadResult.getBuildTimeRunTimeVisibleValues());
             classOutput = Assert.checkNotNullParam("classOutput", builder.getClassOutput());
             roots = Assert.checkNotNullParam("builder.roots", builder.getBuildTimeReadResult().getAllRoots());
-            runTimeDefaults = Assert.checkNotNullParam("runTimeDefaults", builder.getRunTimeDefaults());
             additionalTypes = Assert.checkNotNullParam("additionalTypes", builder.getAdditionalTypes());
             additionalBootstrapConfigSourceProviders = builder.getAdditionalBootstrapConfigSourceProviders();
             staticConfigSources = builder.getStaticConfigSources();
@@ -351,7 +348,7 @@ public final class RunTimeConfigurationGenerator {
                 mc.invokeSpecialMethod(MethodDescriptor.ofConstructor(Object.class), mc.getThis());
                 mc.returnValue(null);
             }
-            if (devMode) {
+            if (liveReloadPossible) {
                 reinit = cc.getMethodCreator(REINIT);
                 reinit.setModifiers(Opcodes.ACC_STATIC | Opcodes.ACC_PUBLIC);
             } else {
@@ -464,7 +461,7 @@ public final class RunTimeConfigurationGenerator {
             // make the build time config global until we read the run time config -
             // at run time (when we're ready) we update the factory and then release the build time config
             installConfiguration(clinitConfig, clinit);
-            if (devMode) {
+            if (liveReloadPossible) {
                 final ResultHandle buildTimeRunTimeDefaultValuesConfigSource = reinit
                         .readStaticField(C_BUILD_TIME_RUN_TIME_DEFAULTS_CONFIG_SOURCE);
                 // create the map for build time config source
@@ -544,7 +541,7 @@ public final class RunTimeConfigurationGenerator {
 
             // create the map for run time specified values config source
             final ResultHandle specifiedRunTimeValues = clinit.newInstance(HM_NEW);
-            if (!devMode) {
+            if (!liveReloadPossible) {
                 //we don't need these in devmode
                 //including it would just cache the first values
                 //but these can already just be read directly, as we are in the same JVM
@@ -553,17 +550,11 @@ public final class RunTimeConfigurationGenerator {
                             clinit.load(entry.getValue()));
                 }
             }
-            for (Map.Entry<String, String> entry : runTimeDefaults.entrySet()) {
-                if (!specifiedRunTimeDefaultValues.containsKey(entry.getKey())) {
-                    // only add entry if the user didn't override it
-                    clinit.invokeVirtualMethod(HM_PUT, specifiedRunTimeValues, clinit.load(entry.getKey()),
-                            clinit.load(entry.getValue()));
-                }
-            }
             final ResultHandle specifiedRunTimeSource = clinit.newInstance(PCS_NEW, specifiedRunTimeValues,
                     clinit.load("Specified default values"), clinit.load(Integer.MIN_VALUE + 100));
+
             cc.getFieldCreator(C_SPECIFIED_RUN_TIME_CONFIG_SOURCE)
-                    .setModifiers(Opcodes.ACC_STATIC | (devMode ? Opcodes.ACC_VOLATILE : Opcodes.ACC_FINAL));
+                    .setModifiers(Opcodes.ACC_STATIC | (liveReloadPossible ? Opcodes.ACC_VOLATILE : Opcodes.ACC_FINAL));
             clinit.writeStaticField(C_SPECIFIED_RUN_TIME_CONFIG_SOURCE, specifiedRunTimeSource);
 
             // add in the custom sources that bootstrap config needs
@@ -709,7 +700,7 @@ public final class RunTimeConfigurationGenerator {
                     // config root field is volatile in dev mode, final otherwise; we initialize it from clinit, and readConfig in dev mode
                     cc.getFieldCreator(rootFieldDescriptor)
                             .setModifiers(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC
-                                    | (devMode ? Opcodes.ACC_VOLATILE : Opcodes.ACC_FINAL));
+                                    | (liveReloadPossible ? Opcodes.ACC_VOLATILE : Opcodes.ACC_FINAL));
 
                     // construct instance in <clinit>
                     ResultHandle instance;
@@ -726,7 +717,7 @@ public final class RunTimeConfigurationGenerator {
                     clinit.invokeVirtualMethod(SB_APPEND_STRING, clinitNameBuilder, clinit.load(root.getName()));
                     clinit.invokeStaticMethod(initGroup, clinitConfig, clinitNameBuilder, instance);
                     clinit.invokeVirtualMethod(SB_SET_LENGTH, clinitNameBuilder, clInitOldLen);
-                    if (devMode) {
+                    if (liveReloadPossible) {
                         instance = readConfig.readStaticField(rootFieldDescriptor);
                         readConfig.invokeVirtualMethod(SB_APPEND_STRING, readConfigNameBuilder,
                                 readConfig.load(root.getName()));
@@ -785,7 +776,7 @@ public final class RunTimeConfigurationGenerator {
             clinit.invokeStaticMethod(CD_UNKNOWN_PROPERTIES,
                     clinit.readStaticField(FieldDescriptor.of(cc.getClassName(), "unused", List.class)));
 
-            if (devMode) {
+            if (liveReloadPossible) {
                 configSweepLoop(siParserBody, readConfig, runTimeConfig, getRegisteredRoots(RUN_TIME));
             }
             // generate sweep for run time
@@ -1672,10 +1663,10 @@ public final class RunTimeConfigurationGenerator {
         }
 
         public static final class Builder {
+            public boolean liveReloadPossible;
             private LaunchMode launchMode;
             private ClassOutput classOutput;
             private BuildTimeConfigurationReader.ReadResult buildTimeReadResult;
-            private Map<String, String> runTimeDefaults;
             private List<Class<?>> additionalTypes;
             private List<String> additionalBootstrapConfigSourceProviders;
 
@@ -1698,21 +1689,17 @@ public final class RunTimeConfigurationGenerator {
                 return this;
             }
 
+            public Builder setLiveReloadPossible(boolean liveReloadPossible) {
+                this.liveReloadPossible = liveReloadPossible;
+                return this;
+            }
+
             BuildTimeConfigurationReader.ReadResult getBuildTimeReadResult() {
                 return buildTimeReadResult;
             }
 
             public Builder setBuildTimeReadResult(final BuildTimeConfigurationReader.ReadResult buildTimeReadResult) {
                 this.buildTimeReadResult = buildTimeReadResult;
-                return this;
-            }
-
-            Map<String, String> getRunTimeDefaults() {
-                return runTimeDefaults;
-            }
-
-            public Builder setRunTimeDefaults(final Map<String, String> runTimeDefaults) {
-                this.runTimeDefaults = runTimeDefaults;
                 return this;
             }
 

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
@@ -4,6 +4,7 @@ import static io.quarkus.deployment.steps.ConfigBuildSteps.SERVICES_PREFIX;
 import static io.quarkus.deployment.util.ServiceUtil.classNamesNamedIn;
 import static java.util.stream.Collectors.toList;
 
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.lang.reflect.Modifier;
 import java.util.ArrayList;
@@ -13,6 +14,7 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Properties;
 import java.util.Set;
 import java.util.stream.Collectors;
 
@@ -32,12 +34,14 @@ import io.quarkus.deployment.builditem.AdditionalStaticInitConfigSourceProviderB
 import io.quarkus.deployment.builditem.ConfigurationBuildItem;
 import io.quarkus.deployment.builditem.ConfigurationTypeBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
+import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
 import io.quarkus.deployment.builditem.LaunchModeBuildItem;
 import io.quarkus.deployment.builditem.LiveReloadBuildItem;
 import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
 import io.quarkus.deployment.builditem.StaticInitConfigSourceFactoryBuildItem;
 import io.quarkus.deployment.builditem.StaticInitConfigSourceProviderBuildItem;
 import io.quarkus.deployment.builditem.SuppressNonRuntimeConfigChangedWarningBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.configuration.BuildTimeConfigurationReader;
 import io.quarkus.deployment.configuration.RunTimeConfigurationGenerator;
@@ -46,15 +50,18 @@ import io.quarkus.deployment.configuration.definition.RootDefinition;
 import io.quarkus.deployment.logging.LoggingSetupBuildItem;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.ClassOutput;
+import io.quarkus.runtime.LaunchMode;
 import io.quarkus.runtime.annotations.ConfigPhase;
 import io.quarkus.runtime.annotations.StaticInitSafe;
-import io.quarkus.runtime.configuration.ConfigChangeRecorder;
+import io.quarkus.runtime.configuration.ConfigRecorder;
+import io.quarkus.runtime.configuration.ConfigUtils;
 import io.quarkus.runtime.configuration.ConfigurationRuntimeConfig;
 import io.quarkus.runtime.configuration.RuntimeOverrideConfigSource;
 import io.smallrye.config.ConfigSourceFactory;
 import io.smallrye.config.PropertiesLocationConfigSourceFactory;
 
 public class ConfigGenerationBuildStep {
+
     @BuildStep
     void deprecatedStaticInitBuildItem(
             List<AdditionalStaticInitConfigSourceProviderBuildItem> additionalStaticInitConfigSourceProviders,
@@ -74,13 +81,27 @@ public class ConfigGenerationBuildStep {
                 PropertiesLocationConfigSourceFactory.class.getName()));
     }
 
+    @BuildStep
+    GeneratedResourceBuildItem runtimeDefaultsConfig(List<RunTimeConfigurationDefaultBuildItem> runTimeDefaults,
+            BuildProducer<NativeImageResourceBuildItem> nativeImageResourceBuildItemBuildProducer)
+            throws IOException {
+        Properties p = new Properties();
+        for (var e : runTimeDefaults) {
+            p.setProperty(e.getKey(), e.getValue());
+        }
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        p.store(out, null);
+        nativeImageResourceBuildItemBuildProducer
+                .produce(new NativeImageResourceBuildItem(ConfigUtils.QUARKUS_RUNTIME_CONFIG_DEFAULTS_PROPERTIES));
+        return new GeneratedResourceBuildItem(ConfigUtils.QUARKUS_RUNTIME_CONFIG_DEFAULTS_PROPERTIES, out.toByteArray());
+    }
+
     /**
      * Generate the Config class that instantiates MP Config and holds all the config objects
      */
     @BuildStep
     void generateConfigClass(
             ConfigurationBuildItem configItem,
-            List<RunTimeConfigurationDefaultBuildItem> runTimeDefaults,
             List<ConfigurationTypeBuildItem> typeItems,
             LaunchModeBuildItem launchModeBuildItem,
             BuildProducer<GeneratedClassBuildItem> generatedClass,
@@ -93,13 +114,6 @@ public class ConfigGenerationBuildStep {
 
         if (liveReloadBuildItem.isLiveReload()) {
             return;
-        }
-
-        Map<String, String> defaults = new HashMap<>();
-        for (RunTimeConfigurationDefaultBuildItem item : runTimeDefaults) {
-            if (defaults.putIfAbsent(item.getKey(), item.getValue()) != null) {
-                throw new IllegalStateException("More than one default value for " + item.getKey() + " was produced");
-            }
         }
 
         Set<String> discoveredConfigSources = discoverService(ConfigSource.class, reflectiveClass);
@@ -120,7 +134,8 @@ public class ConfigGenerationBuildStep {
                 .setBuildTimeReadResult(configItem.getReadResult())
                 .setClassOutput(new GeneratedClassGizmoAdaptor(generatedClass, false))
                 .setLaunchMode(launchModeBuildItem.getLaunchMode())
-                .setRunTimeDefaults(defaults)
+                .setLiveReloadPossible(launchModeBuildItem.getLaunchMode() == LaunchMode.DEVELOPMENT
+                        || launchModeBuildItem.isAuxiliaryApplication())
                 .setAdditionalTypes(typeItems.stream().map(ConfigurationTypeBuildItem::getValueType).collect(toList()))
                 .setAdditionalBootstrapConfigSourceProviders(
                         getAdditionalBootstrapConfigSourceProviders(additionalBootstrapConfigSourceProviders))
@@ -161,7 +176,7 @@ public class ConfigGenerationBuildStep {
     @BuildStep
     @Record(ExecutionTime.RUNTIME_INIT)
     public void checkForBuildTimeConfigChange(
-            ConfigChangeRecorder recorder, ConfigurationBuildItem configItem, LoggingSetupBuildItem loggingSetupBuildItem,
+            ConfigRecorder recorder, ConfigurationBuildItem configItem, LoggingSetupBuildItem loggingSetupBuildItem,
             ConfigurationRuntimeConfig configurationConfig,
             List<SuppressNonRuntimeConfigChangedWarningBuildItem> suppressNonRuntimeConfigChangedWarningItems) {
         BuildTimeConfigurationReader.ReadResult readResult = configItem.getReadResult();

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/DevServicesConfigBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/DevServicesConfigBuildStep.java
@@ -15,11 +15,11 @@ import io.quarkus.deployment.annotations.Produce;
 import io.quarkus.deployment.builditem.DevServicesConfigResultBuildItem;
 import io.quarkus.deployment.builditem.DevServicesLauncherConfigResultBuildItem;
 import io.quarkus.deployment.builditem.DevServicesNativeConfigResultBuildItem;
-import io.quarkus.deployment.builditem.LiveReloadBuildItem;
 import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
 import io.quarkus.deployment.builditem.ServiceStartBuildItem;
 
 class DevServicesConfigBuildStep {
+    static volatile Map<String, String> oldConfig;
 
     @BuildStep
     List<DevServicesConfigResultBuildItem> deprecated(List<DevServicesNativeConfigResultBuildItem> items) {
@@ -30,18 +30,16 @@ class DevServicesConfigBuildStep {
     @BuildStep
     @Produce(ServiceStartBuildItem.class)
     DevServicesLauncherConfigResultBuildItem setup(BuildProducer<RunTimeConfigurationDefaultBuildItem> runtimeConfig,
-            List<DevServicesConfigResultBuildItem> devServicesConfigResultBuildItems,
-            LiveReloadBuildItem liveReloadBuildItem) {
+            List<DevServicesConfigResultBuildItem> devServicesConfigResultBuildItems) {
         Map<String, String> newProperties = new HashMap<>(devServicesConfigResultBuildItems.stream().collect(
                 Collectors.toMap(DevServicesConfigResultBuildItem::getKey, DevServicesConfigResultBuildItem::getValue)));
         Config config = ConfigProvider.getConfig();
-        PreviousConfig oldProperties = liveReloadBuildItem.getContextObject(PreviousConfig.class);
         //check if there are existing already started dev services
         //if there were no changes to the processors they don't produce config
         //so we merge existing config from previous runs
         //we also check the current config, as the dev service may have been disabled by explicit config
-        if (oldProperties != null) {
-            for (Map.Entry<String, String> entry : oldProperties.config.entrySet()) {
+        if (oldConfig != null) {
+            for (Map.Entry<String, String> entry : oldConfig.entrySet()) {
                 if (!newProperties.containsKey(entry.getKey())
                         && config.getOptionalValue(entry.getKey(), String.class).isEmpty()) {
                     newProperties.put(entry.getKey(), entry.getValue());
@@ -51,15 +49,7 @@ class DevServicesConfigBuildStep {
         for (Map.Entry<String, String> entry : newProperties.entrySet()) {
             runtimeConfig.produce(new RunTimeConfigurationDefaultBuildItem(entry.getKey(), entry.getValue()));
         }
-        liveReloadBuildItem.setContextObject(PreviousConfig.class, new PreviousConfig(newProperties));
+        oldConfig = newProperties;
         return new DevServicesLauncherConfigResultBuildItem(Collections.unmodifiableMap(newProperties));
-    }
-
-    static class PreviousConfig {
-        final Map<String, String> config;
-
-        public PreviousConfig(Map<String, String> config) {
-            this.config = config;
-        }
     }
 }

--- a/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigRecorder.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigRecorder.java
@@ -14,9 +14,9 @@ import io.quarkus.runtime.annotations.Recorder;
 import io.quarkus.runtime.configuration.ConfigurationRuntimeConfig.BuildTimeMismatchAtRuntime;
 
 @Recorder
-public class ConfigChangeRecorder {
+public class ConfigRecorder {
 
-    private static final Logger log = Logger.getLogger(ConfigChangeRecorder.class);
+    private static final Logger log = Logger.getLogger(ConfigRecorder.class);
 
     public void handleConfigChange(ConfigurationRuntimeConfig configurationConfig, Map<String, String> buildTimeConfig) {
         Config configProvider = ConfigProvider.getConfig();

--- a/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java
@@ -8,6 +8,7 @@ import static io.smallrye.config.SmallRyeConfig.SMALLRYE_CONFIG_PROFILE_PARENT;
 import static io.smallrye.config.SmallRyeConfigBuilder.META_INF_MICROPROFILE_CONFIG_PROPERTIES;
 
 import java.io.IOException;
+import java.io.InputStream;
 import java.net.URL;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -17,6 +18,7 @@ import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.OptionalInt;
+import java.util.Properties;
 import java.util.Set;
 import java.util.SortedSet;
 import java.util.TreeSet;
@@ -35,6 +37,7 @@ import io.smallrye.config.DotEnvConfigSourceProvider;
 import io.smallrye.config.EnvConfigSource;
 import io.smallrye.config.FallbackConfigSourceInterceptor;
 import io.smallrye.config.Priorities;
+import io.smallrye.config.PropertiesConfigSource;
 import io.smallrye.config.RelocateConfigSourceInterceptor;
 import io.smallrye.config.SmallRyeConfig;
 import io.smallrye.config.SmallRyeConfigBuilder;
@@ -50,6 +53,7 @@ public final class ConfigUtils {
      * The name of the property associated with a random UUID generated at launch time.
      */
     static final String UUID_KEY = "quarkus.uuid";
+    public static final String QUARKUS_RUNTIME_CONFIG_DEFAULTS_PROPERTIES = "quarkus-runtime-config-defaults.properties";
 
     private ConfigUtils() {
     }
@@ -114,6 +118,10 @@ public final class ConfigUtils {
         }
         if (addDiscovered) {
             builder.addDiscoveredSources();
+        }
+        if (runTime || bootstrap) {
+            Map<String, String> runtimeDefaults = loadRuntimeDefaultValues();
+            builder.withSources(new PropertiesConfigSource(runtimeDefaults, "Runtime Defaults", Integer.MIN_VALUE + 50));
         }
         return builder;
     }
@@ -189,14 +197,34 @@ public final class ConfigUtils {
         builder.withSources(provider.getConfigSourceFactory(Thread.currentThread().getContextClassLoader()));
     }
 
+    public static Map<String, String> loadRuntimeDefaultValues() {
+        Map<String, String> values = new HashMap<>();
+        try (InputStream in = Thread.currentThread().getContextClassLoader()
+                .getResourceAsStream(QUARKUS_RUNTIME_CONFIG_DEFAULTS_PROPERTIES)) {
+            if (in == null) {
+                return values;
+            }
+            Properties p = new Properties();
+            p.load(in);
+            for (String k : p.stringPropertyNames()) {
+                if (!values.containsKey(k)) {
+                    values.put(k, p.getProperty(k));
+                }
+            }
+            return values;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
     /**
      * Checks if a property is present in the current Configuration.
-     *
+     * <p>
      * Because the sources may not expose the property directly in {@link ConfigSource#getPropertyNames()}, we cannot
      * reliable determine if the property is present in the properties list. The property needs to be retrieved to make
      * sure it exists. Also, if the value is an expression, we want to ignore expansion, because this is not relevant
      * for the check and the expansion value may not be available at this point.
-     *
+     * <p>
      * It may be interesting to expose such API in SmallRyeConfig directly.
      *
      * @param propertyName the property name.

--- a/extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java
+++ b/extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java
@@ -292,6 +292,13 @@ public class DevServicesDatasourceProcessor {
             compressor.close();
             log.info("Dev Services for " + prettyName
                     + " (" + defaultDbKind.get() + ") started.");
+
+            String devservices = prefix + "devservices.";
+            for (var name : ConfigProvider.getConfig().getPropertyNames()) {
+                if (name.startsWith(devservices)) {
+                    devDebProperties.put(name, ConfigProvider.getConfig().getValue(name, String.class));
+                }
+            }
             return new DevServicesDatasourceResultBuildItem.DbResult(defaultDbKind.get(), devDebProperties);
         } catch (Throwable t) {
             compressor.closeAndDumpCaptured();

--- a/extensions/jdbc/jdbc-postgresql/deployment/pom.xml
+++ b/extensions/jdbc/jdbc-postgresql/deployment/pom.xml
@@ -48,6 +48,16 @@
             <artifactId>assertj-core</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-reactive-deployment</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>rest-assured</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/jdbc/jdbc-postgresql/deployment/src/test/java/io/quarkus/jdbc/postgresql/deployment/DevServicesPostgresqlDatasourceDevModeTestCase.java
+++ b/extensions/jdbc/jdbc-postgresql/deployment/src/test/java/io/quarkus/jdbc/postgresql/deployment/DevServicesPostgresqlDatasourceDevModeTestCase.java
@@ -1,0 +1,53 @@
+package io.quarkus.jdbc.postgresql.deployment;
+
+import java.util.logging.Level;
+
+import javax.inject.Inject;
+
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.agroal.api.AgroalDataSource;
+import io.quarkus.test.QuarkusDevModeTest;
+import io.restassured.RestAssured;
+
+public class DevServicesPostgresqlDatasourceDevModeTestCase {
+
+    @RegisterExtension
+    static QuarkusDevModeTest test = new QuarkusDevModeTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClass(PgResource.class)
+                    .addAsResource(new StringAsset(""), "application.properties"))
+            // Expect no warnings (in particular from Agroal)
+            .setLogRecordPredicate(record -> record.getLevel().intValue() >= Level.WARNING.intValue()
+                    // There are other warnings: JDK8, TestContainers, drivers, ...
+                    // Ignore them: we're only interested in Agroal here.
+                    && record.getMessage().contains("Agroal"));
+
+    @Inject
+    AgroalDataSource dataSource;
+
+    @Test
+    public void testDatasource() throws Exception {
+        RestAssured.get("/pg/save?name=foo&value=bar")
+                .then().statusCode(204);
+
+        RestAssured.get("/pg/get?name=foo")
+                .then().statusCode(200)
+                .body(Matchers.equalTo("bar"));
+
+        test.modifyResourceFile("application.properties", s -> "quarkus.datasource.devservices.properties.log=TRACE");
+
+        RestAssured.get("/pg/get?name=foo")
+                .then().statusCode(404);
+        RestAssured.get("/pg/save?name=foo&value=bar")
+                .then().statusCode(204);
+        RestAssured.get("/pg/get?name=foo")
+                .then().statusCode(200)
+                .body(Matchers.equalTo("bar"));
+    }
+}

--- a/extensions/jdbc/jdbc-postgresql/deployment/src/test/java/io/quarkus/jdbc/postgresql/deployment/PgResource.java
+++ b/extensions/jdbc/jdbc-postgresql/deployment/src/test/java/io/quarkus/jdbc/postgresql/deployment/PgResource.java
@@ -1,0 +1,58 @@
+package io.quarkus.jdbc.postgresql.deployment;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import javax.transaction.Transactional;
+import javax.ws.rs.GET;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.Path;
+import javax.ws.rs.QueryParam;
+
+import io.agroal.api.AgroalDataSource;
+import io.smallrye.common.annotation.Blocking;
+
+@Path("/pg")
+@Blocking
+public class PgResource {
+
+    @Inject
+    AgroalDataSource ds;
+
+    @PostConstruct
+    @Transactional
+    public void setup() throws Exception {
+        try (var con = ds.getConnection()) {
+            try (var smt = con.createStatement()) {
+                smt.executeUpdate("create table foo (name varchar(100) primary key not null, value varchar(100));");
+            }
+        }
+    }
+
+    @GET
+    @Path("save")
+    public void save(@QueryParam("name") String name, @QueryParam("value") String value) throws Exception {
+        try (var con = ds.getConnection()) {
+            try (var smt = con.prepareStatement("insert into foo (name, value) values (?,?)")) {
+                smt.setString(1, name);
+                smt.setString(2, value);
+                smt.execute();
+            }
+        }
+    }
+
+    @GET
+    @Path("get")
+    public String get(@QueryParam("name") String name) throws Exception {
+        try (var con = ds.getConnection()) {
+            try (var smt = con.prepareStatement("select (value) from foo where name = ?")) {
+                smt.setString(1, name);
+                try (var rs = smt.executeQuery()) {
+                    if (!rs.next()) {
+                        throw new NotFoundException();
+                    }
+                    return rs.getString(1);
+                }
+            }
+        }
+    }
+}

--- a/extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java
+++ b/extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java
@@ -30,7 +30,6 @@ import com.mongodb.event.CommandListener;
 import com.mongodb.event.ConnectionPoolListener;
 
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
-import io.quarkus.arc.deployment.BeanArchiveIndexBuildItem;
 import io.quarkus.arc.deployment.BeanContainerBuildItem;
 import io.quarkus.arc.deployment.BeanRegistrationPhaseBuildItem;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
@@ -149,7 +148,7 @@ public class MongoClientProcessor {
     }
 
     @BuildStep
-    public void mongoClientNames(BeanArchiveIndexBuildItem indexBuildItem,
+    public void mongoClientNames(CombinedIndexBuildItem indexBuildItem,
             BuildProducer<MongoClientNameBuildItem> mongoClientName) {
         Set<String> values = new HashSet<>();
         IndexView indexView = indexBuildItem.getIndex();
@@ -203,10 +202,19 @@ public class MongoClientProcessor {
         additionalBeans.produce(AdditionalBeanBuildItem.builder().addBeanClasses(MongoClients.class).setUnremovable().build());
     }
 
+    @BuildStep
+    void connectionNames(
+            List<MongoClientNameBuildItem> mongoClientNames,
+            BuildProducer<MongoConnectionNameBuildItem> mongoConnections) {
+        mongoConnections.produce(new MongoConnectionNameBuildItem(MongoClientBeanUtil.DEFAULT_MONGOCLIENT_NAME));
+        for (MongoClientNameBuildItem bi : mongoClientNames) {
+            mongoConnections.produce(new MongoConnectionNameBuildItem(bi.getName()));
+        }
+    }
+
     @Record(STATIC_INIT)
     @BuildStep
     void build(
-            List<MongoClientNameBuildItem> mongoClientNames,
             MongoClientRecorder recorder,
             SslNativeConfigBuildItem sslNativeConfig,
             CodecProviderBuildItem codecProvider,
@@ -214,7 +222,6 @@ public class MongoClientProcessor {
             BsonDiscriminatorBuildItem bsonDiscriminator,
             CommandListenerBuildItem commandListener,
             List<MongoConnectionPoolListenerBuildItem> connectionPoolListenerProvider,
-            BuildProducer<MongoConnectionNameBuildItem> mongoConnections,
             BuildProducer<SyntheticBeanBuildItem> syntheticBeanBuildItemBuildProducer) {
 
         List<Supplier<ConnectionPoolListener>> poolListenerList = new ArrayList<>(connectionPoolListenerProvider.size());
@@ -230,11 +237,6 @@ public class MongoClientProcessor {
                         bsonDiscriminator.getBsonDiscriminatorClassNames(), commandListener.getCommandListenerClassNames(),
                         poolListenerList, sslNativeConfig.isExplicitlyDisabled()))
                 .done());
-
-        mongoConnections.produce(new MongoConnectionNameBuildItem(MongoClientBeanUtil.DEFAULT_MONGOCLIENT_NAME));
-        for (MongoClientNameBuildItem bi : mongoClientNames) {
-            mongoConnections.produce(new MongoConnectionNameBuildItem(bi.getName()));
-        }
     }
 
     @Record(ExecutionTime.RUNTIME_INIT)

--- a/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/dev/KafkaDevServicesDevModeTestCase.java
+++ b/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/dev/KafkaDevServicesDevModeTestCase.java
@@ -1,0 +1,87 @@
+package io.quarkus.smallrye.reactivemessaging.kafka.deployment.dev;
+
+import java.net.URI;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Supplier;
+
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.sse.SseEventSource;
+
+import org.assertj.core.api.Assertions;
+import org.awaitility.Awaitility;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusDevModeTest;
+import io.quarkus.test.common.http.TestHTTPResource;
+
+public class KafkaDevServicesDevModeTestCase {
+
+    static final String FINAL_APP_PROPERTIES = "mp.messaging.outgoing.generated-price.connector=smallrye-kafka\n" +
+            "mp.messaging.outgoing.generated-price.topic=prices\n" +
+            "mp.messaging.incoming.prices.connector=smallrye-kafka\n" +
+            "mp.messaging.incoming.prices.health-readiness-enabled=false\n" +
+            "mp.messaging.incoming.prices.topic=prices\n";
+
+    @RegisterExtension
+    public static QuarkusDevModeTest test = new QuarkusDevModeTest()
+            .setArchiveProducer(new Supplier<JavaArchive>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(PriceConverter.class, PriceResource.class, PriceGenerator.class)
+                            .addAsResource(new StringAsset(FINAL_APP_PROPERTIES),
+                                    "application.properties");
+                }
+            });
+
+    @TestHTTPResource("/prices/stream")
+    URI uri;
+
+    @Test
+    public void sseStream() {
+        Client client = ClientBuilder.newClient();
+        WebTarget target = client.target(this.uri);
+
+        List<Double> received = new CopyOnWriteArrayList<>();
+
+        try (SseEventSource source = SseEventSource.target(target).build()) {
+            source.register(inboundSseEvent -> received.add(Double.valueOf(inboundSseEvent.readData())));
+            source.open();
+
+            Awaitility.await()
+                    .atMost(Duration.ofSeconds(1))
+                    .until(() -> received.size() >= 2);
+        }
+
+        Assertions.assertThat(received)
+                .hasSizeGreaterThanOrEqualTo(2)
+                .allMatch(value -> (value >= 0) && (value < 100));
+
+        test.modifySourceFile(PriceConverter.class, s -> s.replace("int ", "long "));
+        test.modifySourceFile(PriceGenerator.class,
+                s -> s.replace("Integer", "Long").replace("this.random", "(long)this.random"));
+
+        received.clear();
+
+        try (SseEventSource source = SseEventSource.target(target).build()) {
+            source.register(inboundSseEvent -> received.add(Double.valueOf(inboundSseEvent.readData())));
+            source.open();
+
+            Awaitility.await()
+                    .atMost(Duration.ofSeconds(3))
+                    .until(() -> received.size() >= 2);
+        }
+
+        Assertions.assertThat(received)
+                .hasSizeGreaterThanOrEqualTo(2)
+                .allMatch(value -> (value >= 0) && (value < 100));
+    }
+}

--- a/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/dev/PriceConverter.java
+++ b/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/dev/PriceConverter.java
@@ -1,0 +1,20 @@
+package io.quarkus.smallrye.reactivemessaging.kafka.deployment.dev;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Outgoing;
+
+import io.smallrye.reactive.messaging.annotations.Broadcast;
+
+@ApplicationScoped
+public class PriceConverter {
+    private static final double CONVERSION_RATE = 0.88;
+
+    @Incoming("prices")
+    @Outgoing("processed-prices")
+    @Broadcast
+    public double process(int priceInUsd) {
+        return priceInUsd * CONVERSION_RATE;
+    }
+}

--- a/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/dev/PriceGenerator.java
+++ b/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/dev/PriceGenerator.java
@@ -1,0 +1,22 @@
+package io.quarkus.smallrye.reactivemessaging.kafka.deployment.dev;
+
+import java.time.Duration;
+import java.util.Random;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.reactive.messaging.Outgoing;
+
+import io.smallrye.mutiny.Multi;
+
+@ApplicationScoped
+public class PriceGenerator {
+    private final Random random = new Random();
+
+    @Outgoing("generated-price")
+    public Multi<Integer> generate() {
+        return Multi.createFrom().ticks().every(Duration.ofMillis(10))
+                .onOverflow().drop()
+                .map(tick -> this.random.nextInt(100));
+    }
+}

--- a/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/dev/PriceResource.java
+++ b/extensions/smallrye-reactive-messaging-kafka/deployment/src/test/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/dev/PriceResource.java
@@ -1,0 +1,25 @@
+package io.quarkus.smallrye.reactivemessaging.kafka.deployment.dev;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import org.eclipse.microprofile.reactive.messaging.Channel;
+import org.reactivestreams.Publisher;
+
+@Path("/prices")
+public class PriceResource {
+    private final Publisher<Double> processedPrices;
+
+    public PriceResource(@Channel("processed-prices") Publisher<Double> processedPrices) {
+        this.processedPrices = processedPrices;
+    }
+
+    @GET
+    @Path("/stream")
+    @Produces(MediaType.SERVER_SENT_EVENTS)
+    public Publisher<Double> ssePrices() {
+        return this.processedPrices;
+    }
+}

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
@@ -122,7 +122,7 @@ public class VertxHttpRecorder {
 
     private static volatile Runnable closeTask;
 
-    private static volatile Handler<HttpServerRequest> rootHandler;
+    static volatile Handler<HttpServerRequest> rootHandler;
 
     private static volatile Handler<RoutingContext> nonApplicationRedirectHandler;
 
@@ -1162,6 +1162,14 @@ public class VertxHttpRecorder {
 
     public static Handler<HttpServerRequest> getRootHandler() {
         return ACTUAL_ROOT;
+    }
+
+    /**
+     * used in the live reload handler to make sure the application has not been changed by another source (e.g. reactive
+     * messaging)
+     */
+    public static Object getCurrentApplicationState() {
+        return rootHandler;
     }
 
     public Handler<RoutingContext> createBodyHandler(HttpConfiguration httpConfiguration) {

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/devmode/VertxHttpHotReplacementSetup.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/devmode/VertxHttpHotReplacementSetup.java
@@ -147,11 +147,18 @@ public class VertxHttpHotReplacementSetup implements HotReplacementSetup {
                     synchronized (this) {
                         if (nextUpdate < System.currentTimeMillis() || hotReplacementContext.isTest()) {
                             nextUpdate = System.currentTimeMillis() + HOT_REPLACEMENT_INTERVAL;
+                            Object currentState = VertxHttpRecorder.getCurrentApplicationState();
                             try {
                                 restart = hotReplacementContext.doScan(true);
                             } catch (Exception e) {
                                 event.fail(new IllegalStateException("Unable to perform live reload scanning", e));
                                 return;
+                            }
+                            if (currentState != VertxHttpRecorder.getCurrentApplicationState()) {
+                                //its possible a Kafka message or some other source triggered a reload
+                                //so we could wait for the restart (due to the scan lock)
+                                //but then fail to dispatch to the new application
+                                restart = true;
                             }
                         }
                     }

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -15,6 +15,7 @@ import java.time.Duration;
 import java.time.temporal.ChronoUnit;
 import java.util.AbstractMap;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
@@ -83,6 +84,7 @@ import io.quarkus.test.common.PathTestHelper;
 import io.quarkus.test.common.PropertyTestUtil;
 import io.quarkus.test.common.QuarkusTestResource;
 import io.quarkus.test.common.RestAssuredURLManager;
+import io.quarkus.test.common.RestorableSystemProperties;
 import io.quarkus.test.common.TestClassIndexer;
 import io.quarkus.test.common.TestResourceManager;
 import io.quarkus.test.common.TestScopeManager;
@@ -253,8 +255,8 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
                 hangTaskKey = hangDetectionExecutor.schedule(hangDetectionTask, hangTimeout.toMillis(), TimeUnit.MILLISECONDS);
             }
             ConfigProviderResolver.setInstance(new RunningAppConfigResolver(runningQuarkusApplication));
-
-            System.setProperty("test.url", TestHTTPResourceManager.getUri(runningQuarkusApplication));
+            RestorableSystemProperties restorableSystemProperties = RestorableSystemProperties.setProperties(
+                    Collections.singletonMap("test.url", TestHTTPResourceManager.getUri(runningQuarkusApplication)));
 
             Closeable tm = testResourceManager;
             Closeable shutdownTask = new Closeable() {
@@ -275,6 +277,7 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
                             try {
                                 tm.close();
                             } finally {
+                                restorableSystemProperties.close();
                                 GroovyCacheCleaner.clearGroovyCache();
                                 shutdownHangDetection();
                             }

